lib/widgets/text.dart

```diff
@@ -0,0 +1,120 @@
+///
+/// Some variable fonts can be controlled on a "wght" axis.
+/// Use this to set a value on that axis. It uses [TextStyle.fontVariations],
+/// along with a [TextStyle.fontWeight] that approximates the given "wght"
+/// for the sake of glyphs that need to be rendered by a fallback font
+/// (which might not offer a "wght" axis).
+///
+/// Use this even to specify normal-weight text, by omitting `wght`; then,
+/// [FontWeight.normal.value] will be used. No other layer applies a default,
+/// so if you don't use this, you may e.g. get the font's lightest weight.
+///
+/// Pass [context] to respect a platform request to draw bold text for
+/// accessibility (see [MediaQueryData.boldText]). This handles that request by
+/// using [wghtPlatformRequestsBold] or if that's null, [FontWeight.bold.value].
+///
+/// Example:
+///
+/// ```dart
+/// someTextStyle.merge(weightVariableTextStyle(context, wght: 250)
+/// ```
+///
+/// See also [FontVariation] for more background on variable fonts.
+// TODO(a11y) make `context` required when callers can adapt?
+TextStyle weightVariableTextStyle(BuildContext? context, {
+  int? wght,
+  int? wghtPlatformRequestsBold,
+}) {
+  int value = wght ?? FontWeight.normal.value;
+  if (context != null && MediaQuery.of(context).boldText) {
+    // The framework has a condition on [MediaQueryData.boldText]
+    // in the [Text] widget, but that only affects `fontWeight`.
+    // [Text] doesn't know where to land on the chosen font's "wght" axis if any,
+    // and indeed it doesn't seem updated to be aware of variable fonts at all.
+    value = wghtPlatformRequestsBold ?? FontWeight.bold.value;
+  }
+  assert(value >= 1 && value <= 1000); // https://fonts.google.com/variablefonts#axis-definitions
+
+  return TextStyle(
+    fontVariations: [FontVariation('wght', value.toDouble())],
+
+    // This use of `fontWeight` shouldn't affect glyphs in the preferred,
+    // "wght"-axis font. If it does, see for debugging:
+    //   https://github.com/zulip/zulip-flutter/issues/65#issuecomment-1550666764
+    fontWeight: clampVariableFontWeight(value),
+
+    inherit: true);
+}
+
+/// Find the nearest [FontWeight] constant for a variable-font "wght"-axis value.
+///
+/// Use this for a reasonable [TextStyle.fontWeight] for glyphs that need to be
+/// rendered by a fallback font that doesn't have a "wght" axis.
+///
+/// See also [FontVariation] for background on variable fonts.
+FontWeight clampVariableFontWeight(int wght) {
+  const values = FontWeight.values;
+  assert((() { // `values` is non-empty and sorted
+    FontWeight? last;
+    return values.isNotEmpty && values.every((fontWeight) {
+      final result = (last == null ||
+        (fontWeight.index > last!.index && fontWeight.value > last!.value));
+      last = fontWeight;
+      return result;
+    });
+  })());
+
+  FontWeight? lower;
+  FontWeight? upper;
+  for (final fontWeight in values) {
+    if (fontWeight.value > wght) {
```

Comment:
I was thinking about being agnostic to what the list was (except that it's non-empty and sorted), in case it changes in a new Flutter version. But I guess it's not super likely to change, and also we can have a test or an `assert` to alert us if it changes.